eZ/Publish/Core/Persistence/Cache/Adapter/TransactionalInMemoryCacheAdapter.php

```diff
@@ -0,0 +1,301 @@
+<?php
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+declare(strict_types=1);
+
+namespace eZ\Publish\Core\Persistence\Cache\Adapter;
+
+use Symfony\Component\Cache\Adapter\TagAwareAdapterInterface;
+use Psr\Cache\CacheItemInterface;
+use Symfony\Component\Cache\CacheItem;
+
+/**
+ * Internal proxy adapter invalidating our isolated in-memory cache, and defer shared pool changes during transactions.
+ *
+ * @intenral For type hinting inside eZ\Publish\Core\Persistence\Cache\*. For external, type hint on TagAwareAdapterInterface.
+ *
+ * @todo Naming? Is this something like IsolatedCacheDecorator
+ */
+class TransactionalInMemoryCacheAdapter implements TransactionAwareAdapterInterface
+{
+    /** @var \Symfony\Component\Cache\Adapter\TagAwareAdapterInterface */
+    protected $sharedPool;
+
+    /** @var \eZ\Publish\Core\Persistence\Cache\inMemory\InMemoryCache[] */
+    private $inMemoryPools;
+
+    /** @var int */
+    protected $transactionDepth;
+
+    /** @var array To be unique and simplify lookup hash key is cache tag, value is only true value */
+    protected $deferredTagsInvalidation;
+
+    /** @var array To be unique and simplify lookup hash key is cache key, value is only true value */
+    protected $deferredItemsDeletion;
+
+    /**
+     * @param \Symfony\Component\Cache\Adapter\TagAwareAdapterInterface $sharedPool
+     * @param \eZ\Publish\Core\Persistence\Cache\inMemory\InMemoryCache[] $inMemoryPools
+     * @param int $transactionDepth
+     * @param array $deferredTagsInvalidation
+     * @param array $deferredItemsDeletion
+     */
+    public function __construct(
+        TagAwareAdapterInterface $sharedPool,
+        iterable $inMemoryPools,
+        int $transactionDepth = 0,
+        array $deferredTagsInvalidation = [],
+        array $deferredItemsDeletion = []
+    ) {
+        $this->sharedPool = $sharedPool;
+        $this->inMemoryPools = $inMemoryPools;
+        $this->transactionDepth = $transactionDepth;
+        $this->deferredTagsInvalidation = empty($deferredTagsInvalidation) ? [] : \array_fill_keys($deferredTagsInvalidation, true);
+        $this->deferredItemsDeletion = empty($deferredItemsDeletion) ? [] : \array_fill_keys($deferredItemsDeletion, true);
+        // To modify protected $isHit when items are a "miss" based on deferred delete/invalidation during transactions
+        $this->setCacheItemAsMiss = \Closure::bind(
+            static function (CacheItem $item) {
+                // ... Might not work for anything but new items
+                $item->isHit = false;
```

Comment:
Uh, this one gave me a headache. Too bad Symfony does this exactly like that. Not to mention that PhpStorm shows this as erroneous...
No action required, just complaining about quality of Sf...